### ISSUE TITLE
Version synchronicity between Webapi / Salsah / Sipi / Knora-Base Ontology

### ISSUE THREAD

**Issue by subotic** [dhlab-basel/Knora#513]
We discussed (or raised the question) how are we going to make sure that the correct versions of the different components are used together.

I think that it would help if we would have a way in each component to define the required (min, max, exact) version of the other components.

Then on startup of the `WEBAPI`, we can check if the `knora-base` ontology and `SIPI` have the required version. The same could then be done by `Salsah` and `SIPI`.

In `WEBAPI` (we should really give it a name), we could have a route, that would provide all the versions of all deployed components, which could also help in debugging issues.

**Comment by benjamingeer** [comment]
We can configure the `Server:` header in HTTP responses, to give the versions of the Knora API server and of Sipi. (In Akka HTTP, I think this is done using the [respondWithHeader](http://doc.akka.io/docs/akka-http/10.0.7/scala/http/routing-dsl/directives/respond-with-directives/respondWithHeader.html#respondwithheader) directive). Then when the API server makes a request to Sipi, it can check this value, and when Salsah makes requests to the API server and to Sipi, it can also check what versions they are.

The Knora API server's `Server:` header could include the version of the `knora-base` ontology being used, which could be stored in the triplestore.

When the API server starts, it can also check the version of the `knora-base` ontology and see if it's compatible.

Each component could have configuration options saying the minimum required versions of the other components.

When we do a release, the version number could somehow be automatically stored in a file, which could be read by the software when it starts.

**Comment by tobiasschweizer** [comment]
Great that you are taking care of this!

Do you have an idea how we could address the issue of someone that wants to update his Knora version and wants to find out if or how his data has to be transformed in order to comply with the ontologies?

**Comment by loicjaouen** [comment]
the idea we mentioned last wednesday was to promote the code base to a release (fork branch or label) and put a script that takes care of bringing data in compliance (from the previous).

**Comment by tobiasschweizer** [comment]
You mean automatically calling `TransformData`?

How can we ensure data consistency (that nothing breaks during the conversion)?

**Comment by benjamingeer** [comment]
No, it wouldn't be `TransformData`. We were talking about something that would have to be part of the admin API and would use SPARQL to update the triplestore safely, including updating `knora-base` if necessary.

`TransformData` is OK for test data but we will need something more robust for production use.

**Comment by tobiasschweizer** [comment]
Ok, but do you have already a concrete idea what we would need? What's your guess how much time we need to implement this?

**Comment by benjamingeer** [comment]
We haven't given it much thought yet, but I think we can't make it a priority right now.
